File: 240522/서로 다른 BST 개수 세기/number-of-unique-bst.java
import java.util.*;

public class Main { 
    static Scanner sc = new Scanner(System.in);
    static int n = sc.nextInt();

    public static void main(String[] args) {
        if( n == 1 ) {
            System.out.println(1);
            return;
        }

        if( n == 2 ) {
            System.out.println(2);
            return;
        }

        int[] dp = new int[n+1];
        dp[0] = 1; 
        dp[1] = 1;
        dp[2] = 2;

        for(int i = 3; i <=n; i++) {
            for(int j = 0; j <= i-1; j++ ) {
                dp[i] = dp[i] + dp[j] * dp[i-1-j];
            }
        }
        System.out.println(dp[n]);
    }
}

File: 240522/서로 다른 BST 개수 세기/number-of-unique-bst.kt
import java.util.*

val sc = Scanner(System.`in`)
val n = sc.nextInt()

fun main(){
    if(n == 1) {
        println(1)
        return
    }
    if(n == 2) {
        println(2)
        return
    }
    val dp = IntArray(n+1) { 0 }
    dp[0] = 1
    dp[1] = 1
    dp[2] = 2

    for(i in 3 .. n) {
        for(j in 0 .. i-1) {
            dp[i] = dp[i] + dp[j] * dp[i-1-j]
        }
    }

    println(dp[n])
}